Ready Java Lighting

PREMIEREHELL opened the pull request:
the game will recalculate the light levels for versions tested (19w08b) 1934 to 3949 (1.21.Pre2)

PREMIEREHELL committed:
Ready Java Lighting

the game will recalculate the light levels for (19w08b) 1934 <= 1.14 => 1.21

PREMIEREHELL committed:
Ready Java Lighting

the game will recalculate the light levels for (19w08b) 1934 <= 1.14 => 1.21

Fixed a typo, bug [Bug report] #286

Updated

Fixed this issue where it always evaluates to true and adds the tag.
I think its better to set the value to 0 than to remove the tag,

Revert "Fixed a typo, bug [Bug report] #286"

This reverts commit 16b063564d43441aa2d9ed83d0088b85600f84d1.

PREMIEREHELL committed:
Fixed a typo, bug [Bug report] #286

gentlegiantJGC commented on amulet/level/interfaces/chunk/anvil/anvil_na.py:
```diff
@@ -537,7 +537,7 @@ def _encode_ticks(ticks: Dict[BlockCoordinates, Tuple[str, int, int]]) -> ListTa
                                 "x": IntTag(x),
                                 "y": IntTag(y),
-                                "z": IntTag(x),
+                                "z": IntTag(z),
```
I have added this in its own pull request. Please remove.

PREMIEREHELL replied:
I have not been able to figure out how to remove a commit  with out messing things up. i removed the changes

gentlegiantJGC commented on amulet/level/interfaces/chunk/anvil/anvil_1934.py:
```diff
@@ -1,17 +1,51 @@
+            data,
+            self.isLightOn,
+            chunk.misc.get("isLightOn", None),
+        )
```
This code will always write the isLightOn tag and calculate light will always remove it and mark the chunk as changed.
Either only write it here if it evaluates to True
```py
if chunk.misc.get("isLightOn", None):
    self.set_layer_obj(data, self.isLightOn, ByteTag(1))
```
or modify calculate light to to only remove it if it is ByteTag(1).
The tag being undefined and ByteTag(0) are equivalent to the game. If undefined it will default to ByteTag(0)

PREMIEREHELL replied:
Updated

PREMIEREHELL committed:
Revert "Fixed a typo, bug [Bug report] #286"

This reverts commit 16b063564d43441aa2d9ed83d0088b85600f84d1.

PREMIEREHELL committed:
Updated

Fixed this issue where it always evaluates to true and adds the tag.
I think its better to set the value to 0 than to remove the tag,

gentlegiantJGC commented on amulet/level/interfaces/chunk/anvil/anvil_1934.py:
```diff
@@ -1,17 +1,49 @@
+        self, chunk: Chunk, data: ChunkDataType, floor_cy: int, height_cy: int
+    ):
+        if chunk.misc.get("isLightOn", None):
+            self.set_layer_obj(data, self.isLightOn, ByteTag(0))
```
ByteTag(1)

PREMIEREHELL replied:
That would tell the game that it does not need to be recalculated, it needs to be set to 0

PREMIEREHELL replied:
I see what you mean. "only call this once. Store the value."

gentlegiantJGC commented on amulet/level/interfaces/chunk/anvil/anvil_1934.py:
```diff
@@ -1,17 +1,49 @@
+    def _decode_islighton(
+        self, chunk: Chunk, data: ChunkDataType, floor_cy: int, height_cy: int
+    ):
+        if self.get_layer_obj(data, self.isLightOn) == ByteTag(1):
```
only call this once. Store the value.

PREMIEREHELL replied:
This checks if the tag and value exists in the chunk and if the set value is 1, if it is it stores it, then format.py,
 ```py
changed_ |= chunk.misc.get("isLightOn", None) is not None  #true
``` 
if in format.py the above code, use's pop the below code is None and does not do anything,  
I could put the if statement line 36 below, still would work the same.

```py
if chunk.misc.get("isLightOn", None):  # is true
            self.set_layer_obj(data, self.isLightOn, ByteTag(0) ) # sets it to 0 
```
this is according to my tests
maybe
```py
if chunk.misc.get("isLightOn", None) == ByteTag(1):
            self.set_layer_obj(data, self.isLightOn, ByteTag(0))
```
but then chunk change would always return true

PREMIEREHELL replied:
It looks like nothing gets committed to the originally loaded object data, even after a save it still returns true, until you reload the world in amulet. Does make it a little tedious when testing.

gentlegiantJGC replied:
The purpose of the decoder and encoder is to read the data and write it back exactly as it was found.
The lighting changes in the save operation are a separate step.

PREMIEREHELL committed:
This should work

this will update only the chunks that have the isLightOn tag set to 1 ,

PREMIEREHELL committed:
This should work

gentlegiantJGC commented on amulet/api/chunk/chunk.py:
```diff
@@ -387,3 +389,14 @@ def misc(self) -> dict:
         self._misc = misc
+
+    @property
+    def set_light_off(self) -> bool:
```
This feature only exists in Java. I would prefer if this was left in the misc data.
The rewrite gives each format the ability to define its own chunk class so in the future we could have something like this but for now it should be left in misc.
